### PR TITLE
Charm auto manager

### PR DESCRIPTION
**Pull request by aalavandhan**
Automated manager for weth-wampl charm vault. Manager adjusts vault's liquidity perc, based on AMPL's price deviation from the target.

### CODE REVIEW

File: spot-vaults/contracts/WethWamplManager.sol [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+    /// @notice The contract owner.
+    address public owner;
+
+    //--------------------------------------------------------------------------
+    // Modifiers
+
+    modifier onlyOwner() {
+        // solhint-disable-next-line custom-errors
+        require(msg.sender == owner, "Unauthroized caller");
```
Comment (brandoniles) [inline spot-vaults/contracts/WethWamplManager.sol:67]:
```suggestion
        require(msg.sender == owner, "Unauthorized caller");
```

File: spot-vaults/contracts/WethWamplManager.sol [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    function weth() external view returns (address) {
+        return VAULT.token0();
+    }
+
+    /// @return The vault's token1, the WAMPL token.
+    function wampl() public view returns (IWAMPL) {
+        return IWAMPL(VAULT.token1());
+    }
+
+    /// @notice Computes active liquidity percentage baaed on the provided deviation factor.
```
Comment (brandoniles) [inline spot-vaults/contracts/WethWamplManager.sol:165]:
```suggestion
    /// @notice Computes active liquidity percentage based on the provided deviation factor.
```

File: spot-vaults/contracts/WethWamplManager.sol [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        uint256 x2,
+        uint256 y2,
+        uint256 x
+    ) private pure returns (uint256) {
+        // m = dlY/dlX
+        // c = y2 - m . x2
+        // y = m . x + c
+        int256 dlY = SafeCast.toInt256(y2) - SafeCast.toInt256(y1);
+        int256 dlX = SafeCast.toInt256(x2) - SafeCast.toInt256(x1);
+        int256 c = SafeCast.toInt256(y2) - ((SafeCast.toInt256(x2) * dlY) / dlX);
```
Comment (brandoniles) [inline spot-vaults/contracts/WethWamplManager.sol:321]: I know this is reused from the BB, but looking at this again, is this simpler?
`return y1 + (x - x1) * dlY / dlX`

**Review by brandoniles — COMMENTED** [review]
Overall makes sense to me. Made some minor comments, but would love to get this in front of the Charm devs.

**Review by pixelis0x — APPROVED** [review]
Looks good to me, nice approach

**Review by brandoniles — COMMENTED** [review]
LGTM, the only question I have is where the best place for this code to live is.

The current version is pretty specific to a WAMPL/WETH vault, so it seems a bit out of place in the Spot repo. A very similar version could also manage a spot vault in the future, though.

**Comment by aalavandhan** [comment]
This is a temporary home. Plan to move it into a new monorepo for everything AMPL (similar to this one for SPOT)

**Review by brandoniles — APPROVED** [review]

**Pull request merged**